File: C++/1971. Find if Path Exists in Graph.cpp
class Solution {
public:
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) 
    {
        vector<int>adj[n];
        int m=edges.size();
        for(int i=0;i<m;i++)
        {
            int u=edges[i][0];
            int v=edges[i][1];
            adj[u].push_back(v);
            adj[v].push_back(u);
        }
        vector<bool>visited(n,false);
        queue<int>q;
        visited[source]=true;
        q.push(source);
        while(q.empty()==false)
        {
            int u=q.front();
            q.pop();
            for(int x:adj[u])
            {
                if(visited[x]==false)
                {
                    visited[x]=true;
                    q.push(x);
                }
            }
        }
        return visited[destination];
    }
};

File: JAVA/1971. Find if Path Exists in Graph.java
class Solution {
    public boolean validPath(int n, int[][] edges, int source, int destination) {
        UFDS ufds = new UFDS(n);
        for (int[] edge : edges)
        {
            ufds.merge(edge[0], edge[1]);
        }
        int rootSource = ufds.find(source);
        int rootDestination = ufds.find(destination);
        boolean equal = (rootSource == rootDestination);
        return equal;
    }
    static class UFDS
    {
        int[] p;
        public UFDS(int n) {
            p = new int[n];
            for (int i = 0; i < n; i++) {
                p[i] = i;
            }
        }
        public int find(int x) {
            if (x == p[x]) {
                return x;
            }
            return p[x] = find(p[x]);
        }
        public void merge(int x, int y) {
            int lx = find(x);
            int ly = find(y);
            if (lx != ly)
            {
                p[ly] = lx;
            }
        }
    }
}
